neusomatic/python/sequencing_features.py

```diff
@@ -0,0 +1,379 @@
+logging.basicConfig(level=logging.INFO, format=FORMAT)
+logger = logging.getLogger(__name__)
+
+nan = float('nan')
+
+
+def fisher_exact_test(mat):
+    return fisher.pvalue(mat[0][0], mat[0][1], mat[1][0], mat[1][1]).two_tail
+
+def get_read_pos_for_ref_pos(read, ref_pos_s):
+    cigartuples = read.cigartuples
+    pos_r = read.reference_start
+    current_i = 0
+    output = {}
+    while current_i < len(ref_pos_s):
+        if pos_r > ref_pos_s[current_i] or not cigartuples:
+            output[ref_pos_s[current_i]]=[None, None]
+            current_i +=1
+        else:
+            break
+    if current_i >= len(ref_pos_s):
+        return output
+    cigar_aligned = [cigar_aln_match, cigar_seq_match, cigar_seq_mismatch]
+    cigar_s = 1 if cigartuples[0][0] == cigar_soft_clip else 0
+    cigar_e = (len(cigartuples) - 1) if cigartuples[-1][0] == cigar_soft_clip else len(cigartuples)
+    count = pos_q = cigartuples[0][1] if cigar_s == 1 else 0
+    cigar_index = cigar_s
+    for op, length in cigartuples[cigar_s: cigar_e]:
+        is_aligned = op == 0 or op >= 7
+        delta_r = length if (is_aligned or op == cigar_deletion) else 0
+        delta_q = length if (is_aligned or op == cigar_insertion) else 0
+        while current_i < len(ref_pos_s):
+            diff = ref_pos_s[current_i] - pos_r
+            if diff < delta_r:
+                output[ref_pos_s[current_i]]=[count + diff, (pos_q + diff) if delta_q else None]
+                current_i +=1
+            else:
+                break
+        if current_i >= len(ref_pos_s):
+            return output
+        count += max(delta_r, delta_q)
+        pos_r += delta_r
+        pos_q += delta_q
+        cigar_index += 1
+    while current_i < len(ref_pos_s):
+        output[ref_pos_s[current_i]]=[None, None]
+        current_i +=1
+    return output
+
+class ClusterReads:
+    def __init__(self, bam, variants):
+        self.variants = variants
+        self.chrom = variants[0][0]
+        self.min_pos = variants[0][1]
+        self.max_pos = variants[-1][1]
+        self.reads = []
+        for read_i in bam.fetch(self.chrom, self.min_pos - 1, self.max_pos):
+            if read_i.is_unmapped or not dedup_test(read_i) or read_i.seq is None:
+                continue
+            self.reads.append(read_i)
+
+        done_i = -1
+        n = len(variants)
+        self.var_reads = [[] for i in range(len(variants))]
+        self.read_vars = [[] for i in range(len(self.reads))]
+        for i, read in enumerate(self.reads):
+            for j in range(done_i + 1, n):
+                pos = variants[j][1]
+                if read.reference_start >= pos:
+                    done_i += 1
+                    continue
+                if pos <= read.reference_end:
+                    self.var_reads[j].append(i)
+                    self.read_vars[i].append(j)
+        unused_reads = set(range(len(self.reads)))-set([i for j in self.var_reads for i in j])
```

Comment:
You can use an array of size `len(self.reads)`, say `read_unused_flags` which indicates if a read_index is used or not. Then set membership is just an index into this array and fast.